Add sort functionality

SN9NV opened the pull request:
Adds support for sorting the leaves and pre-images, the reason is because this is what is expected by https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/MerkleProof.sol#L48

Adding this parameter unfortunately changes the API, if you can think of another way to do it, we can give it a try.

I tested the hashed values manually, going through each step in a python script. So I hope it's right.

mcdee commented:
Adding parameters to the existing `NewUsing` will break too much stuff.  I have pushed a commit to provide a parameterized call `NewTree`, please could you rebase and add this as a parameter so that it retains backwards compatibility?  Thanks.

SN9NV commented:
Pretty cool way to solve the problem. Maybe it can also be applied to the `VerifyMultiProofUsing` because I also had to change that interface and maybe it's a good idea in general since that function has a lot of parameters. Just a thought.

mcdee commented:
I have added a `NewMultiProof()` for parameterized multiproof creation, so this should allow you to use that rather than altering existing interfaces.

SN9NV commented:
I have updated the code to use your new parameters. Everything seems to be working now and all tests are passing.

I also added more tests to have 100% coverage.

mcdee commented:
Looking good, please could you address the linting issues and we'll be good to merge?  Thanks.

SN9NV commented:
I believe the lint errors should be solved now.
